Support for Flipping Text in coord_polar()

BradyAJohnston opened the pull request:
This isn't ready for merge yet (things are working but messy), but I've added support for flipping text when the _theta_ of the text is `90 < theta < 270`.   If the option `flip = TRUE` then text will be flipped if would otherwise be written upside down. 

## Regular behaviour:
```r
library(ggplot2)
library(ggfittext)

ggplot(gold, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, 
                 fill = linenumber, label = line)) +
  coord_polar() +
  geom_rect() +
  geom_fit_text(min.size = 0,
                grow = TRUE) +
  scale_fill_gradient(low = "#fee391", high = "#238443")
```
![image](https://user-images.githubusercontent.com/36021261/117235593-9a5c6180-ae59-11eb-8afc-c7ed37f8aed4.png)


## New behaviour with `flip = TRUE`
```r
library(ggplot2)
library(ggfittext)

ggplot(gold, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, 
                 fill = linenumber, label = line)) +
  coord_polar() +
  geom_rect() +
  geom_fit_text(min.size = 0,
                flip = TRUE,
                grow = TRUE) +
  scale_fill_gradient(low = "#fee391", high = "#238443")

```
![image](https://user-images.githubusercontent.com/36021261/117235614-a47e6000-ae59-11eb-8b57-dfc8ae5f2d70.png)

I thought I might try and get your feedback on the feature first before I go in and finish cleaning things up. 

Also thanks for the great package!

BradyAJohnston committed:
test supporting 180 rotation

BradyAJohnston committed:
test add angle

BradyAJohnston committed:
change angle positions

BradyAJohnston committed:
remove theta-90 rot

BradyAJohnston committed:
re-add angle for rot = theta

BradyAJohnston committed:
change hjust

BradyAJohnston committed:
change negative to vjust instead of hjust

BradyAJohnston committed:
adjust vjust

BradyAJohnston committed:
put in actual handling

BradyAJohnston committed:
trying new refactor

BradyAJohnston committed:
test flip

BradyAJohnston committed:
put in flip condition

BradyAJohnston committed:
working flip control

wilkox commented:
This is great! Thanks so much for the contribution and I'll be happy to merge it once you feel it's ready.

A few suggestions:

- Please add yourself as a contributor in DESCRIPTION (`role = "ctb"`) so you get credit!
- Please add a a test or two for the new feature, including a visual test with vdiffr (if you're not familiar with vdiffr just add some tests near the end of tests/testthat/test-polar.R following the `expect_doppelganger()` pattern)
- Currently, the theta value used to decide whether or not to flip is based on the text anchor, which varies based on the value of `place`. However, it seems like `place` is not really relevant to whether or not the text should be flipped; what matters more is whether the majority of the text area is above or below the midline. It might be better to always use the mean of `xmin` and `xmax` to calculate theta i.e. keep lines 218-220 of polar.R, and get rid of the other conditionals based on the value of `place`

Out of curiosity, was there a specific use case that inspired this feature?

BradyAJohnston commented:
Thanks for the feedback! I will clean things up and get and implement those changes and see how we go. 

I've made this specifically for generating plasmid maps in `{ggplot2}` (https://github.com/BradyAJohnston/plasmapR).
<div align="center">
<img src="https://user-images.githubusercontent.com/36021261/118230678-a3d07400-b4c0-11eb-86f1-6b802aded713.png" width="70%">
</div>

I couldn't find anything out there to make ciruclar gene / plasmid maps so I set out to make my own. I looked into your `{gggenes}` which sort of did the trick, but didn't survive `coord_polar()` unfortunately. I was also wanting to implement a `.gb` parser to parse plasmid files as I was also getting errors from other packages that I tried. 

Similarly, I couldn't find a way to do circular text through extensive googling, so I made my own (ended up being implemented the same way that you have done so here, just less elegantly and not in the form of a `geom_*`). 

It wasn't until I spent some time looking again, that I found that `{ggfittext}` also had support for curved text (that actually resized with the plot, unlike my hacky-method) so I started to dig in and implement the flipping to make the plasmid maps more readable :)

Basically it seems like I did a whole bunch of work that you had already done earlier, but I now understand the inner-workings a bit better than previously so I suppose that's a plus!

BradyAJohnston committed:
Add contributor

BradyAJohnston commented:
Hi @wilkox I've had a poke around with what you suggested, but I think leaving it as-is is best. If I change the lines that you suggested, then the positional flipping doesn't work. If left as-is, the positional flipping works correctly. 

The examples below show four of the possible positional arguments, and how the `flip = TRUE` affects them. 

The position is decided by the position argument, and then the text is just flipped in place.

```r 
library(ggplot2)
library(ggfittext)


plotfun <- function(pos, flip) {
  ggplot(gold,
         aes(
           xmin = xmin,
           xmax = xmax,
           ymin = ymin,
           ymax = ymax,
           fill = linenumber,
           label = line
         )) +
    coord_polar(start = 1.3) +
    geom_rect() +
    geom_fit_text(min.size = 0,
                  # grow = TRUE,
                  place = pos,
                  flip = flip) +
    scale_fill_gradient(low = "#fee391", high = "#238443") +
    theme_void() +
    theme(legend.position = "")
}

pos <- c("topright", "topleft", "bottomleft", "bottomright")


non_flip_list <- lapply(pos, plotfun, flip = FALSE)

p1 <- patchwork::wrap_plots(non_flip_list) + 
  patchwork::plot_annotation(tag_levels = list(pos)) + 
    patchwork::plot_annotation(title = "NOT Flipped")

p1
```
### Not-flipped Text
![image](https://user-images.githubusercontent.com/36021261/119486504-e4ff3880-bd8a-11eb-93ae-a57a3218a145.png)


```r
flip_list <- lapply(pos, plotfun, flip = TRUE)

p2 <- patchwork::wrap_plots(flip_list) + 
  patchwork::plot_annotation(tag_levels = list(pos)) + 
  patchwork::plot_annotation(title = "FLIPPED Text")

p2
```

### Flipped Text
![image](https://user-images.githubusercontent.com/36021261/119486435-d022a500-bd8a-11eb-903f-03d7b1d1fb50.png)


I have made some changes and will push them soon and ask for final review :)

BradyAJohnston committed:
add tests for flip featue

BradyAJohnston committed:
add fpip to README

BradyAJohnston committed:
mino documentation for strrev utility

BradyAJohnston commented:
I have now added myself as contributor, added some tests and also added an example to the README. Let me know what you think of my previous post around the `place` values, and if I should still alter more of the code.

wilkox commented on R/geom_fit_text.R:
```diff
@@ -92,6 +92,8 @@
 #' that aesthetics specifying the box must be provided. See Details.
 #' @param data,stat,position,na.rm,show.legend,inherit.aes,... Standard geom
 #' arguments as for `ggplot2::geom_text()`.
+#' @param flip If `TRUE` text will be rotated when it would otherwise be
```
Suggest 'flipped 180 degrees' or 'flipped the right way up' instead of 'rotated'

wilkox commented on README.Rmd:
````diff
@@ -184,14 +184,21 @@ Text can be drawn in polar coordinates with `geom_fit_text()` simply by adding
+p + geom_fit_text(min.size = 0, grow = TRUE)
+```
+
+### Auto-flip text for readability
````
Suggest removing the sub-sub-heading (line 196) and replacing with a descriptive line:

> When text is drawn in polar coordinates, the `flip = TRUE` argument can be used to flip upside-down text the right way up:

wilkox commented:
Thanks again for your work Brady! I agree with your conclusion re. which value to use for theta.

I've made a couple of minor suggested edits to the documentation. If you're happy with those this should be ready for merge.

I suspected you might have been working on something like plasmid mapping! As you may have already guessed, I added the polar coordinate support to ggfittext as a first step towards supporting circular genomes in gggenes, but still haven't found the time to make arrow grobs that work in polar coordinates. If you're agreeable, I'll add a link to plasmapR in the gggenes README in the meantime.

BradyAJohnston committed:
Minor documentation changes

BradyAJohnston commented:
I have incorporated those changes, thanks for the feedback! Happy for you to include the link, I took inspiration from your `{gggenes}` arrows for making mine, I just didn't incorporate them into actual geoms. I'll have to do some more work on the `{plasmapR}` if people are going to be looking. I might even have a look at the gggeenes and poke around at the arrows to see what what I can see.

wilkox commented:
Merged. Thanks!
